### PR TITLE
[UIE-77] add airbnb eslint and prettier

### PR DESCRIPTION
**Pull request by ghost**
### What
- We are moving our style system to AirBnB's Eslint rules, a more standard and well-updated set of rules for Js/Ts formatting. Prettier will also be added to shape the code more consistently.

### Why
- To increase the ease of onboarding, and catch issues we would have otherwise not been aware of.

### Testing strategy
- Not applicable, but tests were run after the style was applied to make sure nothing was broken

### Key Files to Review
- .`eslintrc.js`
- `package.json`
- `tsconfig.json`
- `.eslintignore`

More info here: https://docs.google.com/document/d/1Pi11vy_sjJXahG5o-qHfec3mEQd0q7F2jJqaApV1Bvw/edit#

### CODE REVIEW

**Review by n8landolt — APPROVED** [review]
PUSH THE BUTTON

**Review by Ghost-in-a-Jar — APPROVED** [review]

**Comment by ghost** [comment]
Here goes!

**Comment by dmohs** [comment]
Initial thoughts after a few minutes of reading:

1. This is not a "use AirBnB's linting rules" PR. It is a "use double quotes and semicolons" PR. By size, the change to use AirBnB's style is insignificant. If this were my PR, I would break it into three: (i) change the quotes, (ii) add semicolons, (iii) add linting rules to ensure these changes stick.
2. [AirBnB's style guide](https://github.com/airbnb/javascript#strings) says to "use single quotes." Deviating from that style deserves its own justification. You could put that here or in the suggested formatting-only PR.
3. You should learn about [ignoring commits in git blame](https://www.stefanjudis.com/today-i-learned/how-to-exclude-commits-from-git-blame/) and explain in this or a formatting PR how you will help preserve the usefulness of this feature.

Reading AirBnB's style guide suggests to me that it is a good choice. Each rule gets substantial justification, which is, in my opinion, the most important part of a style guide.

**Comment by cahrens** [comment]
That's interesting about quotes-- I was assuming double quotes (and semicolons) were part of the airbnb standard. @trock-broadinst or @n8landolt  can you say more about how you came to the decision to standardize on double quotes and semicolons? When I reviewed the diff I also found myself wondering if we really wanted to do those 2 specific things, since they result in a huge number of changes.

**Comment by msilva-broad** [comment]
I second the comment on double-quotes and semi-colons.... I am thinking that removing those changes would make this PR quite easier to consume and review.

**Review by aherbst-broad — COMMENTED** [review]
Agreed it would be good to try the alternate quoting method to reduce the churn.

**Comment by n8landolt** [comment]
No opinion on these specific changes here, just my 2 cents.
Making tweaks to these rules before this PR merges will be adding more work for @trock-broadinst - getting this branch in when others aren't merging is difficult.
I suggest any further rule changes be brought up through the UIWG. We have a process to preserve git history, so we should be able to make these changes with minimal impact.
Also, please consider that most of these default rules are used in industry and we should have a strong case for changing them when doing so.

**Comment by nawatts** [comment]
> getting this branch in when others aren't merging is difficult.
> I suggest any further rule changes be brought up through the UIWG. We have a process to preserve git history, so we should be able to make these changes with minimal impact.

Because PRs like this that change nearly every file are pretty much guaranteed to conflict with other PRs, changing the formatting rules is somewhat disruptive and thus there is an advantage to getting them "right" the first time and not changing them again later.

Also, I think a key goal of this change (adding a formatter and stricter ESLint rule) was to, as much as possible, end debate about formatting. I think that if we leave the door open to more sweeping formatting changes, this change itself becomes less valuable.

> Also, please consider that most of these default rules are used in industry and we should have a strong case for changing them when doing so.

I'll claim that a general argument for no semicolons and single quotes can be made based on the first entry in the [coding philosophy](https://github.com/DataBiosphere/terra-ui/wiki/Contributor-Guide#our-coding-philosophy) in the Terra UI wiki:
> Write simple code: clear, concise and easy to understand

Semicolons are (in most cases) unnecessary syntax that don't make the code easier to understand. Removing them makes the code simpler and more concise. In a similar vein, single quotes are visually simpler and more concise than double.

Also, I think "if it isn't broken, don't fix it" is a valid argument. Terra UI devs are used to a certain style and, when in doubt, I think it's preferable to err on the side of "no change".

**Comment by msilva-broad** [comment]
ok, taking a quick look at the airbnb style guide, it looks like:
- simi-colons: yes
- double-quotes: no (use single quotes)

which puts us in the middle.... I would say follow airbnb and move forward:
- keep semi-colons additions in the PR
- switch back to single-quotes in this PR

**Comment by msilva-broad** [comment]
some justification:

Why? When JavaScript encounters a line break without a semicolon, it uses a set of rules called Automatic Semicolon Insertion to determine whether it should regard that line break as the end of a statement, and (as the name implies) place a semicolon into your code before the line break if it thinks so. ASI contains a few eccentric behaviors, though, and your code will break if JavaScript misinterprets your line break. These rules will become more complicated as new features become a part of JavaScript. Explicitly terminating your statements and configuring your linter to catch missing semicolons will help prevent you from encountering issues.

File: package.json
Comment (nawatts) [inline package.json:106]: What is this new dependency for?

**Comment by msilva-broad** [comment]
and I will second Nick's sentiment.  I have a strong bias to accept a generally reasonable set of standards and move forward to the world where we accept it and get the time wins from the tooling auto-formatting.  These types of PRs are hard to not make big, and so are clumsy.  My main concern is that we have quick followthru on folks confirming that the tooling setup is working well.

**Comment by ghost** [comment]
ok, I fixed the quotes, but not the semis, thank you for your research Miguel!

File: package.json
Comment (ghost) [reply package.json:106]: it's required by one of the formatters. It errors out otherwise. This is likely because we use es5 rather than ES2020, according to this post: https://futurestud.io/tutorials/eslint-how-to-fix-bigint-is-not-defined

File: package.json
Comment (nawatts) [inline package.json]: I think it would be preferable to use a version selector that at least specifies a major version instead of "latest" (similar to other dependencies in this file).

Comment (msilva-broad) [reply package.json]: good eyes.  @trock-broadinst can we make sure the PR description above has callouts to the key config/setup files to review?

Comment (ghost) [reply package.json]: added, and fixed the others too, any more nits before the next push? hoping it'll be my last

**Comment by msilva-broad** [comment]
> ok, I fixed the quotes, but not the semis, thank you for your research Miguel!

credit also goes to @n8landolt who dug up the semi-quote justification bits.

File: .eslintrc.js
Comment (msilva-broad) [inline .eslintrc.js]: looks like double-quotes in this file still need to be reverted to single-quotes, which should clean up the diff too.

**Comment by trock-broadinst** [comment]
Alright, we should be ready for final review

File: .eslintrc.js
Comment (ghost) [reply .eslintrc.js]: are there any more files you'd like me to do this on?(e.g. all other root .js files?)

**Review by msilva-broad — APPROVED** [review]
great work on this!

Things like this are always scary.  But auto-formatting is a HUGE win!  Also nice diligence on gitblame fix, etc.

**Comment by cahrens** [comment]
Is the trailing comma on the last item in a list part of the airBnB standard?

**Comment by ghost** [comment]
> Is the trailing comma on the last item in a list part of the airBnB standard?

https://github.com/airbnb/javascript#commas--dangling

File: .eslintrc.js
Comment (cahrens) [inline .eslintrc.js]: @trock-broadinst so shouldn't "comma-dangle" be set to "always" here? 

And meta-question: how can we know that our eslint rules are in sync with the prettier formatting?

Comment (ghost) [reply .eslintrc.js]: I remember there was a complaint about it before, I'm not sure who from, but you can know the rules are in sync because other than the explicitly rewired ones, all eslint and prettier definitions are declared in `extends: ['airbnb-typescript-prettier'],`

Comment (cahrens) [reply .eslintrc.js]: But isn't this an example of them _not_ being in sync? We requiring trailing commas via prettier, but here the value is set to "off".

Comment (ghost) [reply .eslintrc.js]: that's true, I'll fix it, it has to be enabled/disabled in both, so any rules that will auto-format must be applied to both the checker(eslint) and formatter(prettier)

Comment (cahrens) [reply .eslintrc.js]: Thank you! And can you make sure that the other eslint settings are in sync with the prettier settings?

Comment (nawatts) [reply .eslintrc.js]: With the Prettier ESLint plugin and its rule, ESLint requires that code be formatted with Prettier.

https://prettier.io/docs/en/integrating-with-linters.html

Other ESLint formatting rules that overlap with Prettier (such as `max-len`, `comma-dangle`, etc) can either be configured to match Prettier or be disabled, since the Prettier rule enforces formatting on its own.

File: .eslintrc.js
Comment (nawatts) [inline .eslintrc.js:2]: The TypeScript rules included here aren't sourced from AirBnB's style guide, they're from this other 3rd party package, which adds a few rules on top of `@typescript-eslint/recommended`.

Thoughts on depending directly on AirBnB's package and adding TS rules in our own configuration?

Comment (ghost) [reply .eslintrc.js:2]: I'd prefer sticking with this, only because it's used(visibly) by over a thousand other repositories on github, and the main goal here was simply to have a standard to follow that we won't need to maintain ourselves.

Comment (msilva-broad) [reply .eslintrc.js:2]: I can get behind @trock-broadinst 's thinking on this one :)

Comment (nawatts) [reply .eslintrc.js:2]: To be clear, I'm not suggesting configure individual TypeScript related rules. I'm asking about extending from `@typescript-eslint/recommended` and `airbnb` directly, and cutting out the middle man of `eslint-config-airbnb-typescript-prettier`, which is someone's personal project. The argument for basing on AirBnB's style guide as a widely adopted standard doesn't hold as much when we're basing on someone else's config that itself is based on AirBnB's and has a fraction of the adoption of AirBnB's.

Comment (ghost) [reply .eslintrc.js:2]: We've already taken quite a bit of time doing this, and I think what we have now suffices, and migrating would be fairly simple. Can we just merge what we have now before any conflicts arise(please)

Comment (nawatts) [reply .eslintrc.js:2]: I'm not going to request changes over this.

**Review by cahrens — APPROVED** [review]

File: .eslintrc.js
Comment (cahrens) [reply .eslintrc.js:2]: @trock-broadinst you are also going to need a reviewer from data-explorer-eng (CODEOWNERS requirement) in order to merge.

**Review by petesantos — COMMENTED** [review]
One thing to note, when running `yarn start` on this PR for the first time it takes on the order of minutes (on an m1 mac) to start the local dev server. It looks like this is a one time cost, as subsequent starts seem to be fast.

Clicking through the site, there seemed to be some slowness. I am not sure if it is this PR or something else. I am looking to compare to the `dev` branch

**Comment by msilva-broad** [comment]
> One thing to note, when running `yarn start` on this PR for the first time it takes on the order of minutes (on an m1 mac) to start the local dev server. It looks like this is a one time cost, as subsequent starts seem to be fast.
> 
> Clicking through the site, there seemed to be some slowness. I am not sure if it is this PR or something else. I am looking to compare to the `dev` branch

do you have a rough time figure on that first yarn start?

What are other folks seeing?

**Comment by nawatts** [comment]
I assume this change invalidates all Babel / webpack caches and that's why it takes longer on first run.

**Comment by msilva-broad** [comment]
> I assume this change invalidates all Babel / webpack caches and that's why it takes longer on first run.

This is my guess as well... the hot-reload logics going insane most likely.  Pete says the speed concerns go away after that.  But good if we can get a couple more local tests and then warn folks on what to expect.

**Review by s-rubenstein — APPROVED** [review]
I am willing to say this can go in, but I think it is worth talking at retro about making sure that we aren’t rushing past PR comments, because I’ve noticed this isn’t the first time that’s happened

**Comment by dmohs** [comment]
FYI to all: "Auto-merge enabled" seems a bit hazardous for this particular PR.

**Review by petesantos — APPROVED** [review]

**Comment by petesantos** [comment]
> do you have a rough time figure on that first yarn start?

I didn't time it exactly, but I would say on the order of 2-3 minutes on my mac m1

**Comment by ghost** [comment]
is there some way to force merge this, it appears it touches so much code that it's not passing sonarcloud

**Review by cahrens — CHANGES_REQUESTED** [review]
It looks like there was a bad merge with yarn.lock-- I'm seeing that the updated version of webpack is getting removed again.

image

**Review by cahrens — APPROVED** [review]
We will reapply the webpack change afterwards.

File: .eslintrc.js
Comment (pshapiro4broad) [reply .eslintrc.js:2]: @s-rubenstein from `data-explorer-eng` has reviewed this PR

**Pull request merged**
